### PR TITLE
add gitlab imp

### PR DESCRIPTION
**Pull request by letusfly85**
I tried to gitlab imp,

however, something wrong... and it doesn't work.

Now, I'm checking gitlab oauth settings.

### CODE REVIEW

File: pkg/remote/gitlab/gitlab.go [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+func (r *Gitlab) Status(u *common.User, repo *common.Repo, b *common.Build) error {
+	// how can I get Gitlab status
+	return nil
+}
+
+// Netrc returns a .netrc file that can be used to clone
+// private repositories from a remote system.
+func (r *Gitlab) Netrc(u *common.User) (*common.Netrc, error) {
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go:148]: if i correctly understood, this needed for clone via token, but gitlab does not support this /cc @bradrydzewski

File: pkg/remote/gitlab/gitlab.go [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+// GetScript fetches the build script (.drone.yml) from the remote
+// repository and returns in string format.
+func (r *Gitlab) Script(user *common.User, repo *common.Repo, build *common.Build) ([]byte, error) {
+	var client = NewClient(r.URL, user.Token, r.SkipVerify)
+	var path = ns(repo.Owner, repo.Name)
+	return client.RepoRawFile(path, build.Commit.Sha, ".drone.yml")
+}
+
+func (r *Gitlab) Status(u *common.User, repo *common.Repo, b *common.Build) error {
+	// how can I get Gitlab status
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: Currently gitlab doesn't support status for commits and events, also if we want get MR status in gitlab we need implement a special plugin for gitlab, gitlab uses API to fetch build status on client side. But for now we skip this.

File: pkg/remote/gitlab/gitlab.go [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+	var parsed, err = gogitlab.ParseHook(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(parsed.After) == 0 || parsed.TotalCommitsCount == 0 {
+		return nil, nil
+	}
+
+	if parsed.ObjectKind == "merge_request" {
```
Comment (Bugagazavr) [inline pkg/remote/gitlab/gitlab.go]: In drone 0.4 we store source, know about source and target projects https://github.com/drone/drone/pull/634/files

File: pkg/remote/gitlab/gitlab.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+// return default scope for GitHub
+func (r *Gitlab) Scope() string {
+	return DefaultScope
+}
+
+// getStatus is a helper functin that converts a Drone
+// status to a GitHub status.
+func getStatus(status string) string {
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: This can be removed

File: pkg/remote/gitlab/helper.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	return fmt.Sprintf("%s%%2F%s", owner, name)
+}
+
+func GetUserEmail(client *gogitlab.Gitlab, defaultURL string) (*gogitlab.Gitlab, error) {
+	return client, nil
+}
+
+// GetRandom is a helper function that generates a 32-bit random
+// key, base32 encoded as a string value.
+func GetRandom() string {
```
Comment (kzaitsev) [inline pkg/remote/gitlab/helper.go]: Can be removed

File: pkg/remote/gitlab/gitlab.go [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+// GetHost returns the hostname of this remote GitHub instance.
+func (r *Gitlab) GetHost() string {
+	uri, _ := url.Parse(r.URL)
+	return uri.Host
+}
+
+// getDesc is a helper function that generates a description
+// message for the build based on the status.
+func getDesc(status string) string {
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: Can be removed

File: pkg/remote/gitlab/gitlab.go [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+	"github.com/drone/drone/pkg/remote"
+	common "github.com/drone/drone/pkg/types"
+	"github.com/drone/drone/pkg/utils/httputil"
+)
+
+const (
+	DefaultScope = "repo"
+)
+
+const (
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: Can be removed

File: pkg/remote/gitlab/gitlab.go [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+	StatusPending = "pending"
+	StatusSuccess = "success"
+	StatusFailure = "failure"
+	StatusError   = "error"
+)
+
+const (
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: Can be removed

File: pkg/remote/gitlab/gitlab.go [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+	case common.StateFailure:
+		return StatusFailure
+	case common.StateError, common.StateKilled:
+		return StatusError
+	default:
+		return StatusError
+	}
+}
+
+func New(conf *config.Config) *Gitlab {
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: Can be removed, we use `NewDriver` func

File: pkg/remote/gitlab/gitlab.go (hunk 1 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go:148]: thank you, I added NOTE.

File: pkg/remote/gitlab/gitlab.go (hunk 2 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, I added NOTE,
and I want to also check and think about the special plugin for gitlab.

File: pkg/remote/gitlab/gitlab.go (hunk 3 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, I modified this one.

File: pkg/remote/gitlab/gitlab.go (hunk 4 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, I removed.

File: pkg/remote/gitlab/helper.go (hunk 5 above)
Comment (letusfly85) [inline pkg/remote/gitlab/helper.go]: thank you, I removed.

File: pkg/remote/gitlab/gitlab.go (hunk 6 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, I removed.

File: pkg/remote/gitlab/gitlab.go (hunk 7 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, I removed.

File: pkg/remote/gitlab/gitlab.go (hunk 8 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, I removed.

File: pkg/remote/gitlab/gitlab.go (hunk 9 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, I removed.

File: pkg/remote/gitlab/gitlab.go [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+	}
+
+	var hook = new(common.Hook)
+
+	hook.Repo.Name = obj.Source.Name
+	hook.Repo.Owner = obj.Source.Namespace
+
+	// Check pull request comes from public fork
+	if obj.Source.VisibilityLevel < 20 {
+		//hook.SourceRemote = obj.Source.SshUrl
```
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: It seems that the new data model doesn't have `SourceRemote`...
What should I use?

File: pkg/remote/gitlab/gitlab.go [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+		}
+	} else {
+		//hook.SourceRemote = obj.Source.HttpUrl
+	}
+
+	hook.Commit.Author.Login = req.FormValue("owner")
+	//hook.Repo = req.FormValue("name")
+	hook.Commit.Sha = obj.LastCommit.Id
+	hook.Commit.Branch = obj.TargetBranch
+	//hook.Commit.SourceBranch = obj.SourceBranch
```
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: It seems that the new data model doesn't have `SourceBranch`...
What should I use?

File: pkg/remote/gitlab/gitlab.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+		// check for non-internal pull request
+		if obj.Source.Name != obj.Target.Name || obj.Source.Namespace != obj.Target.Namespace {
+			return nil, nil
+		}
+	} else {
+		//hook.SourceRemote = obj.Source.HttpUrl
+	}
+
+	hook.Commit.Author.Login = req.FormValue("owner")
+	//hook.Repo = req.FormValue("name")
```
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: It seems that the new data model doesn't have `Repo`...
What should I use?

**Comment by letusfly85** [comment]
I modified, please see it.

Thank you.

File: pkg/remote/gitlab/gitlab.go (hunk 10 above)
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: Ah, sorry, looks like drone API changed again, need to remove pull request support, in gitlab 8.0 ( september release ) gitlab got same MR model as github https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/981

Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: @Bugagazavr 

All right, thank you!

> gitlab got same MR model as github

So, the hook method also can be same as `github.go`?

Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: I think, currently we need return `nil` if we got pull request from gitlab, when this feature has been implemented in gitlab we can add support for pull requests, for now i have no idea how we can do this.

Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: model changes are O.K, I will see.

I also added NOTE for hook method, for gitlab 8.0 release on September.

Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: I modified for returning `nil` when we got pull request from gitlab, also add note for it.

please see it, thank you.

File: pkg/remote/gitlab/gitlab.go [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+		return nil, nil
+	}
+
+	var hook = new(common.Hook)
+
+	hook.Repo.Name = obj.Source.Name
+	hook.Repo.Owner = obj.Source.Namespace
+
+	// Check pull request comes from public fork
+	if obj.Source.VisibilityLevel < 20 {
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: This also must be removed because it's a part of `MR` support

File: pkg/remote/gitlab/gitlab.go [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+	}
+
+	obj := parsed.ObjectAttributes
+	if !(obj.State == "opened" && obj.MergeStatus == "unchecked") {
+		return nil, nil
+	}
+
+	var hook = new(common.Hook)
+
+	hook.Repo.Name = obj.Source.Name
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: This must be removed, because this a part of `MR` too

File: pkg/remote/gitlab/gitlab.go [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+	hook.Repo.Owner = obj.Source.Namespace
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: And this (:

File: pkg/remote/gitlab/gitlab.go [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+	var parsed, _ = gogitlab.ParseHook(payload)
+
+	if parsed.ObjectKind == "merge_request" {
+		// NOTE: in gitlab 8.0, gitlab will get same MR models as github
+		//       https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/981/diffs
+		return nil, nil
+	}
+
+	obj := parsed.ObjectAttributes
+	if !(obj.State == "opened" && obj.MergeStatus == "unchecked") {
```
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: And remove this too

File: pkg/remote/gitlab/gitlab.go (hunk 10 above)
Comment (kzaitsev) [inline pkg/remote/gitlab/gitlab.go]: if you need better example look at this https://github.com/Bugagazavr/drone/blob/gitlab_pr/plugin/remote/gitlab/gitlab.go#L151-L187

For now you have many legacy from `MR` method

Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thanks, I see it and modified hook method.

File: pkg/remote/gitlab/gitlab.go (hunk 15 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: I see, thank you.

File: pkg/remote/gitlab/gitlab.go (hunk 16 above)
Comment (letusfly85) [inline pkg/remote/gitlab/gitlab.go]: thank you, removed.

**Comment by kzaitsev** [comment]
Ok, thanks

**Comment by letusfly85** [comment]
Thank you!

**Comment by letusfly85** [comment]
@Bugagazavr 

I'm so sorry...... I forgot modifying `Godeps.json` or `Makefile` after `go get` libraries.

However I don't know about `godep` with `_workspace` :sweat_drops: 

How can I `godep` ....?

**Comment by kzaitsev** [comment]
i fix it

**Comment by letusfly85** [comment]
Thank you, from next time, I will be care about it.

**Pull request merged**